rs/boundary_node/anonymization/client/src/lib.rs

```diff
@@ -1 +1,253 @@
+pub enum RegisterError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Register: Sync + Send {
+    async fn register(&self, pubkey: &[u8]) -> Result<(), RegisterError>;
+}
+
+/// LeaderMode indicates whether a new salt is required
+#[derive(Debug)]
+pub enum LeaderMode {
+    /// Generate a fresh salt
+    Bootstrap,
+
+    /// Refresh the encrypted values
+    Refresh,
+}
+
+impl From<ifc::LeaderMode> for LeaderMode {
+    fn from(value: ifc::LeaderMode) -> Self {
+        match value {
+            ifc::LeaderMode::Bootstrap => LeaderMode::Bootstrap,
+            ifc::LeaderMode::Refresh => LeaderMode::Refresh,
+        }
+    }
+}
+
+/// Pair associates a principal with a binary-blob (pubkey, ciphertext, etc)
+pub struct Pair(pub Principal, pub Vec<u8>);
+
+impl std::fmt::Debug for Pair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Pair")
+            .field(&self.0.to_text())
+            .field(&"..")
+            .finish()
+    }
+}
+
+impl From<&ifc::Pair> for Pair {
+    fn from(value: &ifc::Pair) -> Self {
+        Self(value.0, value.1.to_owned())
+    }
+}
+
+impl From<&Pair> for ifc::Pair {
+    fn from(value: &Pair) -> Self {
+        Self(value.0, value.1.to_owned())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum QueryError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error("unavailable")]
+    Unavailable,
+
+    #[error("leader assignment received")]
+    Leader(LeaderMode, Vec<Pair>),
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Query: Sync + Send {
+    async fn query(&self) -> Result<Vec<u8>, QueryError>;
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SubmitError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Submit: Sync + Send {
+    async fn submit(&self, vs: &[Pair]) -> Result<(), SubmitError>;
+}
+
+#[derive(Clone)]
+pub struct Canister {
+    agent: Agent,
+    cid: Principal,
+}
+
+impl Canister {
+    pub fn new(agent: Agent, cid: Principal) -> Self {
+        Self { agent, cid }
+    }
+}
+
+fn nonce() -> Vec<u8> {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_nanos()
+        .to_le_bytes()
+        .to_vec()
+}
+
+#[async_trait]
+impl Register for Canister {
+    async fn register(&self, pubkey: &[u8]) -> Result<(), RegisterError> {
+        use ifc::{RegisterError as Error, RegisterResponse as Response};
+
+        let args = Encode!(&pubkey).context("failed to encode arg")?;
+
+        let cid = self
+            .cid
+            .as_slice()
+            .try_into()
+            .context("failed to convert cid")?;
+
+        let resp = self
+            .agent
+            .execute_update(
```

Comment:
See my comment above regarding `Agent` and unit testing. IMO a trait would be useful here for unit testing.

Reply:
What would we be unit testing though? These are very thin functions that just make use of the underlying agent.

Reply:
> What would we be unit testing though? 

You could test for the expected errors, cause ATM this is not possible, i suppose. It's just a suggestion, if this function remains thin, maybe it's an overkill.

```
#[automock]
#[async_trait]
pub trait ExecutesCalls: Sync + Send {
    async fn execute_update(&self) -> Result<....>;
    async fn execute_query(&self) -> Result<...>;
}

pub struct Canister<Agent> {
    agent: Agent,
}

impl<Agent: ExecutesCalls> Register for Canister<Agent> {
    async fn register(&self) -> Result<(), String> {
        let response = self.agent.execute_update();
        ...
    }
}
```